core/src/com/mygdx/minigolf/controller/EntityFactory.java

```diff
@@ -0,0 +1,48 @@
+package com.mygdx.minigolf.controller;
+
+import com.badlogic.ashley.core.Component;
+import com.badlogic.ashley.core.Engine;
+import com.badlogic.ashley.core.Entity;
+import com.mygdx.minigolf.model.EntityType;
+import com.mygdx.minigolf.model.components.Graphical;
+import com.mygdx.minigolf.model.components.Objective;
+import com.mygdx.minigolf.model.components.Physical;
+import com.mygdx.minigolf.model.components.Player;
+import com.mygdx.minigolf.model.components.PowerUpAffectable;
+import com.mygdx.minigolf.model.components.PowerUpGiver;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+
+public class EntityFactory {
+
+    private final Engine engine;
+
+    Map<EntityType, HashSet<Class<? extends Component>>> componentMap = new HashMap<EntityType, HashSet<Class<? extends Component>>>() {{
+        put(EntityType.GOLFBALL, new HashSet<>(Arrays.asList(Player.class, Physical.class, Graphical.class, PowerUpAffectable.class)));
+        put(EntityType.HOLE, new HashSet<>(Arrays.asList(Physical.class, Graphical.class, Objective.class)));
+        put(EntityType.SPAWN, new HashSet<>(Collections.singletonList(Physical.class)));
+        put(EntityType.POWERUP, new HashSet<>(Arrays.asList(Physical.class, Graphical.class, PowerUpGiver.class)));
+        put(EntityType.DEFAULT, new HashSet<>(Arrays.asList(Physical.class, Graphical.class)));
+    }};
+
+    public EntityFactory(Engine engine) {
+        this.engine = engine;
+    }
+
+    public Entity createEntity(EntityType entityType) {
+        Entity e = new Entity();
+        componentMap.get(entityType).forEach(c -> {
+            try {
+                e.add(c.newInstance());
+            } catch (InstantiationException | IllegalAccessException instantiationException) {
+                instantiationException.printStackTrace();
+            }
+        });
+        engine.addEntity(e);
+        return e;
+    }
```

Comment:
Ser nå at det var et elendig forslag å bruke maps. Hva med å lagre komponentene direkte i enumen? Slik:
```
public enum EntityType {
    GOLFBALL(Player.class, Physical.class, Graphical.class, PowerUpAffectable.class),
    HOLE(Physical.class, Graphical.class, Objective.class),
    SPAWN(Physical.class),
    POWERUP(Physical.class, Graphical.class, PowerUpGiver.class),
    DEFAULT(Physical.class, Graphical.class);

    public final List<Class<?>> components;

    EntityType(Class<?>... components) {
        this.components = Collections.unmodifiableList(Arrays.asList(components));
    }

}
```
og så i EntityFactory:
```
public class EntityFactory {

    private final Engine engine;

    public EntityFactory(Engine engine) {
        this.engine = engine;
    }

    public Entity createEntity(EntityType entityType) {
        Entity entity = new Entity();
        entityType.components.forEach(c -> {
            try {
                entity.add((Component) c.newInstance());
            } catch (InstantiationException | IllegalAccessException e) {
                e.printStackTrace();
            }
        });
        return entity;
    }
}
```

Reply:
Kan være at unmodifiableList er et dårlig valg. Kan alternativt gjøre components privat og så lage en getComponents-metode som returnerer en kopi av components.
Foretrekker personlig å unngå getere/setere der det passer, og det er kanskje ikke så bra å lage masse ekstra kopier av listene?

Reply:
Hva er DEFAULT egt? Obstacle?

Reply:
Obstacle og Surface

Reply:
Okei det gir mening. Kan være det er bedre for lesbarhet å ha de to som egne entityTyper. Altså
```
...
OBSTACLE(Physical.class, Graphical.class),
SURFACE(Physical.class, Graphical.class);
```